app/src/main/java/de/westnordost/streetcomplete/quests/smoking/AddSmoking.kt

```diff
@@ -0,0 +1,57 @@
+package de.westnordost.streetcomplete.quests.smoking
+
+import de.westnordost.streetcomplete.R
+import de.westnordost.streetcomplete.data.osm.mapdata.Element
+import de.westnordost.streetcomplete.data.osm.mapdata.MapDataWithGeometry
+import de.westnordost.streetcomplete.data.osm.mapdata.filter
+import de.westnordost.streetcomplete.data.meta.updateWithCheckDate
+import de.westnordost.streetcomplete.data.osm.osmquests.OsmFilterQuestType
+import de.westnordost.streetcomplete.data.osm.osmquests.Tags
+import de.westnordost.streetcomplete.data.user.achievements.QuestTypeAchievement.CITIZEN
+import de.westnordost.streetcomplete.ktx.arrayOfNotNull
+
+class AddSmoking : OsmFilterQuestType<SmokingAllowed>() {
+
+    private val elementFilterBasic = """
+        nodes, ways with
+        (
+          amenity ~ bar|cafe|pub|biergarten|restaurant|nightclub|stripclub
+          or leisure ~ outdoor_seating
+        )
+    """
+
+    override val elementFilter = elementFilterBasic + """
+	and (!smoking or smoking older today -8 years)
+    """
+
+    override val changesetComment = "Add smoking status"
+    override val wikiLink = "Key:smoking"
+    override val icon = R.drawable.ic_quest_smoke
+    override val isReplaceShopEnabled = true
+    override val defaultDisabledMessage = R.string.default_disabled_msg_go_inside_regional_warning
+
+    override val questTypeAchievements = listOf(CITIZEN)
+
+    override fun getTitle(tags: Map<String, String>) =
+        if (hasFeatureName(tags))
+            R.string.quest_smoking_name_type_title
+        else
+            R.string.quest_smoking_no_name_title
+
+    override fun getTitleArgs(tags: Map<String, String>, featureName: Lazy<String?>): Array<String> {
+        val name = tags["name"] ?: tags["brand"]
+        return arrayOfNotNull(name, featureName.value)
+    }
+
+    override fun getHighlightedElements(element: Element, getMapData: () -> MapDataWithGeometry) =
+        getMapData().filter(elementFilterBasic)
```

Comment:
I think that it is OK to highlight also shops and other shoplike objects, like in other similar quests

Reply:
You mean by using `isKindOfShopExpression()` I've seen some other quest do? I guess I could, but:
- it misses half the tags I need highlighed, so I'd need to extend it (either `isKindOfShopExpression()` itself (which I'm too afraid to do without external guidance), of locally add extra filters. Also then I'd need to watch if `isKindOfShopExpression()` possibly gets modified in a way that requires me to update the filter (e.g. if some of terms get dropped or limited there), and commiting to "eternal maintenance" is a kind of turn-off for me :smiley: 

- Also, I'm not sure if it would really improve the situation? according to the source code comments:
> Elements that should be highlighted on the map alongside the selected one because they provide context for the given element. For example, nearby benches should be shown when answering a question for a bench so the user knows which of the benches is meant.

If I'm reading that correctly, as I'm mapping `smoking/no smoking` places, what I want to accomplish is to avoid is that users mixes up one cafe/pub (where there might be `smoking`) with another. But showing a lot of extra data around (like shops, offices, crafts, banks etc) does not really help much with that in my opinion. Sure, it sometimes perhaps might help somewhat with locating the cafe (_"ok, this cafe is to the left of post office"_), but it also possibly creates a lot of additional (unwanted?) clutter. 

But if you think that would be improvement, I can replace that with e.g.:
```diff
-getMapData().filter(elementFilterBasic)
+getMapData().filter(elementFilterBasic + isKindOfShopExpression() + " or " + isKindOfShopExpression("disused"))
```

If that is the way to go, then it would be good to also mention `isKindOfShopExpression()` in new quest docs.
 
(Or did I completely missed the point here?)

Reply:
Yes, union here works. Showing all items that would be of interest to this quest AND all shoplike stuff seems optimal.

Maybe also elementFilterBasic which are disused (would require it to support prefixes like say in https://github.com/streetcomplete/StreetComplete/commit/972637797e93a1f96dd8ec3da14e348598f8bfa7 )

Reply:
implemented

Reply:
Just my luck; in the meantime `isKindOfShopExpression()` stopped existing in the `master` :smiling_face_with_tear: